elastic/kibana
# [saved objects] Adds bulkDelete API

Resolves https://github.com/elastic/kibana/issues/30503

## Summary

The saved objects service has API's for updating, getting, and resolving many documents in one request but is missing similar functionality for deleting many documents. 

There are a few use cases for a new API that would delete documents in bulk:
 - bulk delete alert saved objects
 - bulk delete telemetry-related documents
 - replacement for clearing the kibana index with `esArchiver`

This PR adds a `bulkDelete` API with the following characteristics:
Multi-namespace saved objects shared to more than one space need to be deleted by force since the requestor in the active space may not be aware of potential repercussions.

### How force in bulk delete behaves:
We implement a force-deletion at a request level:
- when `force` is false (or not added as a query), the document isn't deleted and the API returns an error for that document.
- with `force` as true: will force delete the document.
The requestor must be authorized to delete saved objects in all the spaces from which the object would be removed when using `force`. This applies to all objects requested in the bulk delete operation.

Once a document has been deleted, the API deletes any legacy URL aliases that may have existed for the object.
The behavior differs from that of deleting a single document with the saved objects `delete` API in that `force` is applied to all the objects sent in the request, as opposed to applying to one particular document.

### Other notable differences between `delete` and `bulkDelete`:
**Errors**: 
- `bulkDelete` returns errors in the response for each document that couldn't be deleted, whereas `delete` returns an empty object.
- `bulkDelete` returns a bad request for deleting hidden objects, whereas `delete` throws a not-found error.

**Response**:
- `bulkDelete` responds with an array of `statuses`, containing a summary of the outcome for deleting an object.The response per-object consistd of a `success` status as either `true` or `false` and will always be present. If an object couldn't be deleted (`success: false`), the corresponding status entry also includes the error for that object.

### How to test this:
Testing bulk delete saved objects that don't exist in the current callee's space:
1. start es and start kibana
2. create a new space "Non default"
3. change spaces into "Non default"
4. load all three sample data fixtures
5. open the browser's network tab
6. navigate to Saved Objects application and open the network request for the `_find?perPage=50&page=1&fields...` request
7. copy the contents of the `saved_objects` array from the response
8. Using a tool of choice (I used postman), issue a `_bulk_delete` request to saved objects from the default space:

<details>
<summary> Issue a `_bulk_delete` request to saved objects from the default space </summary>

```
POST localhost:5601/{basepath}/api/saved_objects/_bulk_delete
body (as raw JSON):
[
  {"type": "visualization", "id": "314c6f60-2224-11e8-b802-5bcf64c2cfb4"},
  {"type": "lens", "id": "16b1d7d0-ea71-11eb-8b4b-f7b600de0f7d"},
  ...rest  
]
```

</details>

<details>
<summary> Observe that the request fails with: </summary>

```json
{
    "statuses": [
        {
            "success": false,
            "id": "314c6f60-2224-11e8-b802-5bcf64c2cfb4",
            "type": "visualization",
            "error": {
                "message": "Unable to delete saved object id: 314c6f60-2224-11e8-b802-5bcf64c2cfb4, type: visualization that exists in multiple namespaces, use the \"force\" option to delete all saved objects: Bad Request",
                "statusCode": 400,
                "error": "Bad Request"
            }
        },
        {
            "success": false,
            "id": "16b1d7d0-ea71-11eb-8b4b-f7b600de0f7d",
            "type": "lens",
            "error": {
                "message": "Unable to delete saved object id: 16b1d7d0-ea71-11eb-8b4b-f7b600de0f7d, type: lens that exists in multiple namespaces, use the \"force\" option to delete all saved objects: Bad Request",
                "statusCode": 400,
                "error": "Bad Request"
            }
        }
    ]
}
```

</details>

9.  Send the request again but from the "Non default" space:
`POST localhost:5601/{basepath}/s/non-default/api/saved_objects/_bulk_delete`

<details>
<summary>Observer that the request is successful:</summary>

```json
{
    "statuses": [
        {
            "success": true,
            "id": "d3d7af60-4c81-11e8-b3d7-01146121b73d",
            "type": "index-pattern"
        },
        {
            "success": true,
            "id": "b05ee7f1-0341-43f1-b345-f6cd3a944254",
            "type": "visualization"
        }
    ]
}
```

</details>

#### Testing legacy URL alias deletion
There's no easy way to test that `bulkDelete` also deletes legacy URL aliases that may be present for multi-namespace saved objects.

You can use staging data following the [instructions](https://github.com/elastic/kibana/pull/107099#issuecomment-891147792) for testing `bulkResolve`. I found I had to create a new user a `system_indices_superuser` role add the staging data to kibana.

### Checklist

Delete any items that are not applicable to this PR.


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Code should gracefully handle failures for users with different privileges | High | High | Integration tests will verify that saved objects are not deleted when not permitted to delete them in any situation where the request would violate the privileges for the callee. |
| Consumers need to know the implications of using `force` to delete multi-namespace saved objects | Medium | High | Consumers need to opt-in to using `force` explicitly |
| Performance regression when deleting multi-namespace saved objects with many legacy URL aliases | Medium | High | Consumers need to be aware that deleting multi-namespace objects also deletes legacy URL aliases, which is done one at a time. THere will be a follow up issue to optimise the `legacyUrlAlias` deletion  |
